crates/polars-ops/src/chunked_array/top_k.rs

```diff
@@ -32,46 +42,154 @@ where
+            let values = arg_partition(&mut v, k, descending, TotalOrd::tot_cmp);
             let mut out = ChunkedArray::from_iter(values.iter().copied());
             out.rename(ca.name());
             out
         },
     }
 }
 
+fn top_k_bool_impl(
+    ca: &ChunkedArray<BooleanType>,
+    k: usize,
+    descending: bool,
+) -> ChunkedArray<BooleanType> {
+    if ca.null_count() == 0 {
+        let true_count = ca.sum().unwrap() as usize;
+        let mut bitmap = MutableBitmap::with_capacity(k);
+        if !descending {
+            // true first
+            bitmap.extend_constant(std::cmp::min(k, true_count), true);
+            bitmap.extend_constant(k.saturating_sub(true_count), false);
+        } else {
+            let false_count = ca.len().saturating_sub(true_count);
+            bitmap.extend_constant(std::cmp::min(k, false_count), false);
+            bitmap.extend_constant(k.saturating_sub(false_count), true);
+        }
+        let arr = BooleanArray::from_data_default(bitmap.into(), None);
+        unsafe {
+            ChunkedArray::from_chunks_and_dtype(ca.name(), vec![Box::new(arr)], DataType::Boolean)
+        }
+    } else {
+        let null_count = ca.null_count();
+        let true_count = ca.sum().unwrap() as usize;
+        use std::iter;
+        if !descending {
+            // Null -> True -> False
+            let mut new_ca: BooleanChunked = iter::repeat(None)
+                .take(null_count)
+                .chain(iter::repeat(Some(true)).take(true_count))
+                .chain(iter::repeat(Some(false)).take(ca.len() - true_count - null_count))
+                .take(k)
+                .collect_trusted();
+            new_ca.rename(ca.name());
+            new_ca
+        } else {
+            // False -> True -> Null
+            let mut new_ca: BooleanChunked = iter::repeat(Some(false))
+                .take(null_count)
+                .chain(iter::repeat(Some(true)).take(true_count))
+                .chain(iter::repeat(None).take(ca.len() - true_count - null_count))
+                .take(k)
+                .collect_trusted();
+            new_ca.rename(ca.name());
+            new_ca
+        }
+    }
+}
+
+fn top_k_binary_impl(
+    ca: &ChunkedArray<BinaryType>,
+    k: usize,
+    descending: bool,
+) -> ChunkedArray<BinaryType> {
+    if k >= ca.len() {
+        return ca.sort(!descending);
+    }
+
+    // descending is opposite from sort as top-k returns largest
+    let k = if descending {
+        std::cmp::min(k, ca.len())
+    } else {
+        ca.len().saturating_sub(k + 1)
+    };
+
+    if ca.null_count() == 0 {
+        let mut v: Vec<&[u8]> = Vec::with_capacity(ca.len());
+        for arr in ca.downcast_iter() {
+            v.extend(arr.non_null_values_iter());
+        }
+        let values = arg_partition(&mut v, k, descending, TotalOrd::tot_cmp);
+        ChunkedArray::from_slice(ca.name(), values)
+    } else {
+        let mut v = ca.iter().collect::<Vec<_>>();
```

Comment:
Can you use `downcast_iter` and `extend`? Similar to the non null branch. Flattening the loops is not without cost.